presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -360,20 +568,159 @@ protected Page computeNext()
+        {
+            // wrapper each function with an aggregator
+            ImmutableList.Builder<Aggregator> builder = ImmutableList.builder();
+
+            requireNonNull(accumulatorFactories, "accumulatorFactories is null");
+            for (int i = 0; i < accumulatorFactories.size(); i++) {
+                AccumulatorFactory accumulatorFactory = accumulatorFactories.get(i);
+                builder.add(new Aggregator(accumulatorFactory, step, numberOfColumnsInKey + i));
+            }
+            return builder.build();
+        }
+
+        private void convertToMerge()
+        {
+            step = Step.partialInput(step);
+            ImmutableList.Builder<Integer> groupByPartialChannels = ImmutableList.builder();
+            for (int i = 0; i < groupByTypes.size(); i++) {
+                groupByPartialChannels.add(i);
+            }
+            if (hashChannel.isPresent()) {
+                hashChannel = Optional.of(groupByTypes.size());
+            }
+
+            groupByChannels = groupByPartialChannels.build();
+        }
+
+        private void closeSpiller()
+        {
+            if (spiller.isPresent()) {
+                spiller.get().close();
+                spiller = Optional.empty();
+            }
+        }
+
+        private List<Type> buildIntermediateTypes()
+        {
+            ArrayList<Type> types = new ArrayList<>(groupByHash.getTypes());
+            for (Aggregator aggregator : aggregators) {
+                types.add(aggregator.getIntermediateType());
+            }
+            return types;
+        }
+
+        private List<Type> buildTypes()
+        {
+            ArrayList<Type> types = new ArrayList<>(groupByHash.getTypes());
+            for (Aggregator aggregator : aggregators) {
+                types.add(aggregator.getType());
+            }
+            return types;
+        }
+
+        private long getSizeInMemory()
+        {
+            long sizeInMemory = groupByHash.getEstimatedSize();
+            for (Aggregator aggregator : aggregators) {
+                sizeInMemory += aggregator.getEstimatedSize();
+            }
+            sizeInMemory -= operatorContext.getOperatorPreAllocatedMemory().toBytes();
+            if (sizeInMemory < 0) {
+                sizeInMemory = 0;
+            }
+            return sizeInMemory;
+        }
+
+        private Iterator<Integer> consecutiveGroupIds()
+        {
+            return new Iterator<Integer>()
+            {
+                private final int groupCount = groupByHash.getGroupCount();
+                private int groupId = 0;
+
+                @Override
+                public boolean hasNext()
+                {
+                    return groupId < groupCount;
+                }
+
+                @Override
+                public Integer next()
+                {
+                    return groupId++;
+                }
+            };
+        }
+
+        private Iterator<Integer> sortedGroupIds()
```

Comment:
can't you just return `List` here and in `consecutiveGroupIds` and use `List` in `buildResultFromMemory`? Is there a concern about memory limit?